controlplane/internal/gateway/proxy.go

```diff
@@ -0,0 +1,202 @@
+package gateway
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/siderolabs/grpc-proxy/proxy"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/yanet-platform/yanet2/controlplane/internal/xgrpc"
+)
+
+// TransparentWebGRPCProxy is an HTTP handler that translates HTTP requests
+// with binary protobuf payloads to gRPC calls.
+type TransparentWebGRPCProxy struct {
+	registry *BackendRegistry
+	log      *zap.SugaredLogger
+}
+
+// NewHTTPHandler creates a new HTTPHandler.
+func NewHTTPHandler(registry *BackendRegistry, log *zap.SugaredLogger) *TransparentWebGRPCProxy {
+	return &TransparentWebGRPCProxy{
+		registry: registry,
+		log:      log,
+	}
+}
+
+// ServeHTTP handles HTTP requests.
+func (m *TransparentWebGRPCProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
+	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
+
+	if r.Method == http.MethodOptions {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	// Only support POST method for actual gRPC calls.
+	if r.Method != http.MethodPost {
+		http.Error(w, "Method not allowed. Only POST is supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Convert URL path to gRPC fullMethodName format.
+	fullMethodName := r.URL.Path
+	fullMethodName = strings.TrimPrefix(fullMethodName, "/api")
+	if !strings.HasPrefix(fullMethodName, "/") {
+		fullMethodName = "/" + fullMethodName
+	}
+
+	service, method, err := xgrpc.ParseFullMethod(fullMethodName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Invalid method name: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	backend, ok := m.registry.GetBackend(service)
+	if !ok {
+		http.Error(w, fmt.Sprintf("Service not found: %s", service), http.StatusNotFound)
+		return
+	}
+
+	// Read request body (binary protobuf payload).
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to read request body: %v", err), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	// Create context with metadata from HTTP headers.
+	ctx := r.Context()
+	md := metadata.New(nil)
+	for k, v := range r.Header {
+		k = strings.ToLower(k)
+		if !strings.HasPrefix(k, "sec-") && k != "host" && k != "connection" {
+			md.Set(k, v...)
+		}
+	}
+	ctx = metadata.NewOutgoingContext(ctx, md)
+
+	outCtx, conn, err := backend.GetConnection(ctx, fullMethodName)
+	if err != nil {
+		m.log.Errorw("failed to get connection to backend",
+			zap.String("service", service),
+			zap.String("method", method),
+			zap.Error(err),
+		)
+		http.Error(w, fmt.Sprintf("Failed to connect to backend: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Create byte buffer for response.
+	var respData []byte
```

Comment:
Why was the declaration placed so far from its actual usage?

Reply:
Fixed.